README.md
# stattools
An R package of tools for statistical analyses


NEWS.md
# stattools 0.5.0

* **New** functions `t_test()` and `ks_test()` to perform Student's t-Test and
  Kolmogorov-Smirnov Test on autocorrelated data.
* The addition of these functions makes a license change necessary. `stattools`
  is now distributed under GPL (>= 3).

# stattools 0.4.0

* **New** function `getEffectiveDOF()` to calculate the effective degrees of
  freedom of an autocorrelated dataset.

# stattools 0.3.0

* **New** function `ShiftCorrelation()` to run a shift correlation analysis
  between two data vectors in order to find the optimal shift maximising the
  correlation;
* added unit test infrastructure.

# stattools 0.2.0

Minor changes:

* `rmsd()` now includes the option to calculate the RMSD of one vector against
  zero automatically when only a single vector is passed as first argument.
* Souce files in `R/` have been reorganized.
* R CMD check now passes with zero notes or warnings.
* Code base was beautified to adhere to style guide.
* Function documentations were improved.

# stattools 0.1.0

* Initial package version with number and scope of the functions incl. their
  documentation identical to their original versions in deprecated package
  'ecustools' which they were a part of ('ecustools'
  [`main`](https://github.com/EarthSystemDiagnostics/ecustools/tree/master)
  branch as of 2020-11-20).


tests/testthat.R
library(testthat)
library(stattools)

test_check("stattools")


tests/testthat/test-cor-functions.R
test_that("shift correlation works.", {

  expected1 <- data.frame(i = -2, cor = 1, row.names = as.integer(3))
  expected2 <- data.frame(i = -2, shift = -10, cor = 1,
                          row.names = as.integer(3))

  y1 <- 1 : 4
  y2 <- c(1, 2, 1, 5)

  shifts <- -4 : 1

  expect_error(ShiftCorrelation(y1[1], y2, shifts))
  expect_equal(ShiftCorrelation(y1, y2, shifts), expected1)
  expect_equal(ShiftCorrelation(y1, y2, shifts, unit = 5), expected2)

})


tests/testthat/test-getEffectiveDOF.R
test_that("DOF calculation works", {
  expect_equal(getEffectiveDOF(10, 0.), 10)
  expect_equal(round(getEffectiveDOF(10, 0.6), 6), 3.072636)
})


R/rmsd.R
#' Root-mean-square deviation
#'
#' Calculate the root-mean-square deviation (rmsd) of two numeric vectors.
#'
#' @param v1 numeric vector for which to compute the rmsd with \code{v2}; if
#'   that is \code{NULL} (the default) the rmsd of \code{v1} relative to zero is
#'   calculated.
#' @param v2 numeric vector for which to compute the rmsd with \code{v1}
#'   (optional); if given, it must be of the same length as \code{v1}, if
#'   \code{NULL} (the default) the rmsd of \code{v1} relative to zero is
#'   calculated.
#' @param na.rm a logical value indicating whether \code{NA} values should be
#'   stripped before the computation proceeds. Defaults to \code{FALSE}.
#' @return The root-mean-square deviation of \code{v1} and \code{v2}, or
#'   \code{NA} (for \code{na.rm = FALSE}) if any of their elements is
#'   \code{NA}; or, if \code{v2} is \code{NULL}, the rmsd from zero of
#'   \code{v1}.
#' @author Thomas Münch
#' @examples
#' v1 <- rnorm(1000)
#' v2 <- rnorm(1000)
#' rmsd(v1, v2)
#' @export
rmsd <- function(v1, v2 = NULL, na.rm = FALSE) {

  if (is.null(v2)) v2 <- rep(0, length(v1))

  if (length(v1) != length(v2)) {
    stop("Arguments must have the same length.")
  }
  
  sqrt(mean((v1 - v2)^2, na.rm = na.rm))
}


R/sd-functions.R
#' Population standard deviation
#'
#' Calculate the population standard deviation.
#'
#' @param x numeric; vector of values.
#' @return The population standard deviation value of \code{x}.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#'
#' # analytical solution to SD of uniform distribution:
#' sqrt(1/12 * 10^2)
#'
#' sd(1:10)
#'
#' PopSD(1:10)
#' @export
PopSD <- function(x) {

  n <- length(x)
  adj <- sqrt((n - 1) / n)
  adj * stats::sd(x)
}

#' Standard deviation (error) of standard deviation
#'
#' Calculate the standard deviation (i.e. standard error) of a standard
#' deviation, given n, the number of samples from which the standard deviation
#' was estimated.
#'
#' If n <= 300, an exact method is used, otherwise Stirling's approximation is
#' used to avoid numerical errors calculating the gamma function for large
#' values of n. Formulae taken from https://stats.stackexchange.com/q/28567.
#'
#' @param s numeric; estimated standard deviation.
#' @param n integer; number of samples from which \code{s} was estimated.
#' @return Numeric value giving the standard error of the estimated standard
#'   deviation \code{s}.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#' \dontrun{
#' if(interactive()){
#'  SDofSD(1, 12)
#'  library(ggplot2)
#' df <- expand.grid(SD = 1:10, n = seq(2, 10, length.out = 100))
#' df$SDofSD = SDofSD(df$SD, df$n)
#' ggplot(df, aes(x = n, y = SDofSD, group = SD, colour = SD)) +
#'   geom_line()
#'  }
#' }
#' @export
SDofSD <- function(s, n) {

  if (any(n < 2, na.rm = TRUE))
    warning("n must be greater than or equal to 2")
  if (any(s < 0, na.rm = TRUE))
    warning("s must be >= 0")
  if(length(s) > 1 & length(n) == 1)
    n <- rep(n, length(s))
  if(length(n) > 1 & length(s) == 1)
    s <- rep(s, length(n))
  if(length(s) != length(n))
    stop("'s' and 'n' should be vectors of either the same length, ",
         "or one of them should be length == 1.")

  ifelse(n <= 300, {
    message("Method: Exact")
    g.a <- gamma((n - 1) / 2)
    g.b <- gamma(n / 2)
    s * (g.a / g.b) * sqrt((n - 1) / 2 - (g.b / g.a) ^ 2)
  }, {
    message("Method: Stirling's approximation")
    s * sqrt(exp(1) * (1 - 1 / n) ^ (n - 1) - 1)
  })
}

#' Standard deviation of numerical probability density function
#'
#' Calculates the standard deviation for a given numerical probability density
#' function, i.e. a vector of values and corresponding vector of densities. It
#' is important that the range of x values covers a sufficient fraction of the
#' total density. For continuous distributions it should extend well into the
#' tails. The approximation will be better the finer the resolution of x.
#'
#' @param x numeric; vector of values.
#' @param d numeric; vector of densities, do not need to sum to 1.
#' @return Numeric value giving the standard deviation of the given probability
#'   density function.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#' \dontrun{
#' if(interactive()){
#'  # Example 1
#'  x <- seq(1, 100, by = 0.01)
#'  d <- dnorm(x, 60, 5)
#'  plot(x, d, type = "l")
#'  SDofNumDist(x, d)
#'
#'  # Example 2
#'  a <- 45
#'  b <- 55
#'  d2 <- dunif(x, a, b)
#'  plot(x, d2, type = "l")
#'  SDofNumDist(x, d2)
#'
#'  # analytical solution for uniform distribution
#'  sqrt(1/12 * (b-a)^2)
#'  }
#' }
#' @export
SDofNumDist <- function(x, d) {

  d <- d / sum(d)
  sqrt(sum(d * x^2) - sum(d * x)^2)
}


R/cor-functions.R
#' Lag correlation analysis
#'
#' Find the optimal shift (positive or negative) between two data vectors from
#' maximising the correlation within their common overlap.
#'
#' @param y1 numeric vector of first data series.
#' @param y2 numeric vector of second data series which is shifted relative to
#'   \code{y1}; needs to have the same length as \code{y1}.
#' @param shifts integer vector of shifts (can be positive, negative, or
#'   both). For each value, \code{y2} is shifted the corresponding number of
#'   elements earlier or later, and the correlation with \code{y1} is computed.
#' @param unit optional value for the unit which matches the shifts to the
#'   physical scale on which the data is tabulated.
#' @param plot logical; plot an optional graph of correlation versus tried
#'   shifts.
#' @param xlab optional x axis label for the plot.
#' @param ylab optional y axis label for the plot.
#' @return Data frame with one row with the optimal shift and the corresponding
#'   correlation value; if a \code{unit} is provided, it includes an additional
#'   column with the optimal shift in the real units.
#' @author Thomas Münch
#' @examples
#' y1 <- rnorm(100)
#' y2 <- rnorm(100)
#'
#' ShiftCorrelation(y1, y2, shifts = (-10 : 10))
#'
#' # display optimal shift in the real units of the data:
#' ShiftCorrelation(y1, y2, shifts = (-10 : 10), unit = 3)
#' @export
ShiftCorrelation <- function(y1, y2, shifts, unit = NULL,
                             plot = FALSE, xlab = NULL, ylab = NULL) {

  if (length(y1) != length(y2)) {
    stop("'y1' and 'y2' must have the same length.", call. = FALSE)
  }

  res <- data.frame(
    i = shifts,
    cor = sapply(shifts, function(k) {
      stats::cor(y1, Hmisc::Lag(y2, k), use = "pairwise.complete.obs")})
  )

  optim <- res[which.max(res$cor), ]

  if (plot) {

    pnt <- optim
    if (length(unit)) {
      res$i <- res$i * unit
      pnt$i <- pnt$i * unit
      if (!length(xlab)) xlab <- "Shift in real units"
    }

    if (!length(xlab)) xlab <- "Bin shift"
    if (!length(ylab)) ylab <- "Correlation"

    graphics::plot(res, type = "b",
                   xlab = xlab, ylab = ylab, pch = 19, lwd = 1.5)
    graphics::points(pnt, pch = 19, col = "red", cex = 1.25)

  }

  if (length(unit)) {
    optim$shift <- optim$i * unit
    optim <- optim[, c(1, 3, 2)]
  }

  return(optim)

}


R/getEffectiveDOF.R
#' Effective degrees of freedom
#'
#' Calculate the effective degrees of freedom of an autocorrelated dataset
#' consisting of \code{n} observations, i.e. the nuber of independent
#' obervations given an autocorrelation parameter for the data estimated at lag
#' 1.
#'
#' @param n number of observations.
#' @param a1 lag 1 autocorrelation parameter.
#' @return the effective degrees of freedom.
#' @author Thomas Laepple, Thomas Münch
#' @examples
#' getEffectiveDOF(10, 0.6)
#' @source
#' Xu, X.: Methods in Hypothesis Testing, Markov Chain Monte Carlo and
#'   Neuroimaging Data Analysis, PhD thesis, Harvard University, Cambridge,
#'   Massachusetts, 2013, \url{https://dash.harvard.edu/handle/1/11108711}.
#' @export
#'
getEffectiveDOF <- function(n, a1) {

  a1 <- abs(a1)
  denom <- (1 + a1) / (1 - a1) - (2 * a1 / n) * (1 - a1^n) / (1 - a1)^2

  n / denom

}


R/distribution-functions.R
#' Temperature distribution of seasonal cycle
#'
#' Probability density function and random number generation for the temperature
#' distribution in the seasonal cycle.
#'
#' The seasonal temperature cycle is approximated as a sine wave with gaussian
#' noise. The probability density function is obtained by numerically
#' convolving a scaled sine wave with a normal PDF with mean equal to 0. Random
#' deviates are generated by adding gaussian noise to points sampled uniformly
#' from the scaled sine wave on the interval 0 to 2*pi.
#'
#' @param x numeric; vector of temperatures.
#' @param mean.t numeric; mean annual temperature.
#' @param amp.t numeric; peak-to-peak amplitude of the seasonal cycle in
#'   temperature.
#' @param sd.t numierc; inter-annual standard deviation of the temperature at a
#'   given point in the seasonal cycle.
#' @param return return either the densities for the temperatures \code{x}, or a
#'   function to do so; default: c("density", "FUN").
#' @param res numeric; resolution of the temperature vectors to be convolved;
#'   default is 0.01.
#' @param n integer; number of samples.
#' @name SeasonalCycle
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#' \dontrun{
#' if(interactive()){
#' amp.t <- 10
#' mean.t <- 25
#' sd.t <- 0.75
#'
#' x <- seq(mean.t - amp.t, mean.t + amp.t, length.out = 1000)
#' Z <- rSeas(n = 100000, mean.t, amp.t, sd.t)
#'
#' hist(Z, freq=F,breaks=100)
#' lines(x, dSeas(x, mean.t, amp.t, sd.t, res = 0.1), col = "Blue")
#' lines(x, dSeas(x, mean.t, amp.t, sd.t, res = 0.01), col = "Green")
#' lines(x, dSeas(x, mean.t, amp.t, sd.t, res = 0.001), col = "Red")
#'  }
#'}
NULL

#' @rdname SeasonalCycle
#' @export
dSeas <- function(x = NULL, mean.t, amp.t, sd.t,
                  return = c("density", "FUN"), res = 0.01) {

  rng <- max(c(amp.t, sd.t * 5))
  lnth <- (2 * rng) / res

  # make vector length with many factors for fast FFT
  lnth2 <- (2 ^ ceiling(log(lnth, base = 2)))

  if (lnth2 > 5000) {
    warning("Sequences to be convolved are very long, ",
            "decrease the resolution or rescale 'mean.t' and 'amp.t'.")
  }

  # calculate new resolution for scaling
  res2 <- (2 * rng) / lnth2

  xin <- seq(mean.t - rng, mean.t + rng, length.out = lnth2)
  zin <- dSin(xin, mean.t, amp.t)

  yin <- stats::dnorm(xin, mean.t, sd.t)

  cv <- stats::convolve(zin, yin, type = "open")
  cv <- cv / (sum(cv) * res2)

  x2 <- seq(mean.t - 2 * rng, mean.t + 2 * rng, length.out = length(cv))

  FUN <- stats::approxfun(x2, cv, yleft = 0, yright = 0)

  ret <- match.arg(return)

  switch(ret,
         density = return(FUN(x)),
         FUN = return(FUN))
}

#' @rdname SeasonalCycle
#' @export
rSeas <- function(n, mean.t, amp.t, sd.t){
  x <- stats::runif(n, 0, 2*pi)
  y <- mean.t + sin(x) * amp.t/2
  z <- y + stats::rnorm(n, 0, sd.t)
  return(z)
}

#' Probability density of scaled sine wave
#'
#' Calculate the probability density for a scaled sine wave.
#'
#' @param x numeric; vector of values for which to obtain the probability
#'   densities.
#' @param mean.t numeric; mean of the sine wave.
#' @param amp.t numeric; peak-to-peak amplitude of the sine wave.
#' @return Numeric vector of probability density values for the values given in
#'   \code{x}. Values in \code{x} outside the sine wave's envelope obtain zero
#'   probability.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @export
dSin <- function(x, mean.t, amp.t) {

  d <- rep(0, length(x))

  # Do not evaluate where known to be NaN
  ind <- x > (mean.t - amp.t / 2) & x < (mean.t + amp.t / 2)
  d[ind] <- 1 / (pi * sqrt(1 - ((2 * x[ind]) / amp.t - (2 * mean.t) / amp.t) ^ 2))

  2 * d / amp.t
}

#' Summarise empirical distribution
#'
#' Summarise an empirical probability distribution function.
#'
#' Calculation of the mode is naive. For a multimodal distribution only the
#' highest is returned, in the case of 2 or more modes with exactly the same
#' probability, the first is returned.
#'
#' @param x numeric; vector of values of the empirical PDF.
#' @param p numeric; vector of probabilities of the empirical PDF.
#' @return Returns a named vector with the mean, median, mode, and standard
#' deviation of the empirical PDF.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#' df <- data.frame(x = 1:10)
#' df$p <- dnorm(df$x, 5, 2)
#' SummariseEmpiricalPDF(df$x, df$p)
#' @export
SummariseEmpiricalPDF <- function(x, p) {

  # Ensure x and p are sorted
  p <- p[order(x)]
  x <- sort(x)

  # Ensure p sum to 1
  p <- p / sum(p)

  # Mean
  w.mean <- sum(x * p)

  # SD
  M <- sum(p > 0)
  w.sd <- sqrt(sum(p * (x - w.mean)^2) / ((M - 1)/ M * sum(p)))

  # Median
  csum.p <- cumsum(p)
  med.ind <- which.min(abs(csum.p - 0.5))
  w.median <- x[med.ind]

  # Mode
  max.wt <- max(p)
  n.max <- sum(p == max.wt)
  if (n.max > 1) {
    warning(paste0(n.max,
                   " x with equal maximum probability. Returning the first"))
  }
  mode <- x[which.max(p)]

  return(c("mean" = w.mean, "median" = w.median, "mode" = mode, "sd" = w.sd))
}

#' Expected range of observations
#'
#' Calculates the expected range of n observations drawn from a distribution with
#' standard deviation sd.
#'
#' @param sd numeric; standard deviation.
#' @param n integer; number of observations.
#' @return Numeric value with the expected range of the observation values,
#'   i.e. the difference between maximum and minimum value.
#' @author Andrew Dolman <andrew.dolman@awi.de>
#' @examples
#' ExpectedRange(5, 15)
#'
#' \dontrun{
#' library(tidyverse)
#' f <- function(x, n) n * x * pnorm(x)^(n - 1) * dnorm(x)
#' ests <- expand.grid(mean = 0, sd = 2, n = 15, rep = 1:1000, i = 1:15) %>%
#'   mutate(rdv = rnorm(n(), mean, sd)) %>%
#'   group_by(mean, sd, n, rep) %>%
#'   summarise(sd_hat = sd(rdv),
#'             obs.range = diff(range(rdv))) %>%
#'   group_by(mean, sd, n, rep) %>%
#'   mutate(sd.range = 2*sd_hat*integrate(f,-Inf,Inf, n = n)$value,
#'          exp.range = ExpectedRange(sd, n))
#' ests %>%
#'   gather(estimate, value, -mean, -sd, -n, -rep, -sd_hat, -exp.range) %>%
#'   ggplot(aes(x = estimate, y = value)) +
#'   geom_boxplot() +
#'   geom_hline(aes(yintercept = exp.range)) +
#'   facet_wrap(~n)
#'   }
#' @export
ExpectedRange <- function(sd, n) {

  f <- function(x, n) {n * x * stats::pnorm(x)^(n - 1) * stats::dnorm(x)}

  exp.range <- 2 * sd * stats::integrate(f, -Inf, Inf, n = n)$value

  return(exp.range)
}


R/stattest-functions.R
# Code adapted from base stats:::t.test.default as of R version 4.3.1:
# GPL-2 | GPL-3
# Copyright (C) 2023 R Core Team

# Changes to original code by Thomas Muench in 2023/11/27:
# - introduced function parameters `a1.x` and `a1.y` to allow input of AR1
#   autocorrelation parameters for data x and/or y;
# - use these parameters to reset the number of x/y data points to the effective
#   number of degrees of freedom using stattools::getEffectiveDOF

#' Student's t-Test for autocorrelated data
#'
#' Perform a one-sample or a two-sample t-test accounting for autocorrelation of
#' the input data. This function is a modified version of the base R's default
#' \code{\link[stats]{t.test}}, which treats autocorrelation of the input data
#' by resetting the number of observations to the effective degrees of freedom
#' according to user-supplied estimates of the data's autocorrelation parameters
#' at lag one. Note that this code is adapted from the default method of the
#' t.test function, thus it cannot be used with class \code{formula}.
#'
#' @param x a (non-empty) numeric vector of data values.
#' @param y an optional (non-empty) numeric vector of data values.
#' @param a1.x autocorrelation parameter at lag one of the data in \code{x}.
#' @param a1.y autocorrelation parameter at lag one of the data in \code{y}.
#' @param alternative a character string specifying the alternative hypothesis,
#'   must be one of \code{"two.sided"} (default), \code{"greater"} or
#'   \code{"less"}.  You can specify just the initial letter.
#' @param mu a number indicating the true value of the mean (or difference in
#'   means if you are performing a two sample test).
#' @param paired a logical indicating whether you want a paired t-test.
#' @param var.equal a logical variable indicating whether to treat the two
#'   variances as being equal. If ‘TRUE’ then the pooled variance is used to
#'   estimate the variance otherwise the Welch (or Satterthwaite) approximation
#'   to the degrees of freedom is used.
#' @param conf.level confidence level of the interval.
#' @param ... further arguments to be passed to or from methods.
#' @return see \code{?t.test}.
#' @seealso \code{\link[stats]{t.test}}, \code{\link{getEffectiveDOF}}
#' @source \code{stats:::t.test.default}
#' @author R Core Team, Thomas Münch
#' @examples
#'
#' # Check the performance of the test on autocorrelated data
#'
#' # This function Monte Carlo samples autocorrelated data series from an AR1
#' # process and runs the default t test as well as the stattools t_test to
#' # estimate the tests' rejection rates, i.e. the relative number of false
#' # positives where the tests reject the null hypothesis
#' estimateRejectionRate <- function(nobs, nr, a1.x, a1.y = a1.x,
#'                                   method = c("one.sample", "two.sample")) {
#'
#'  method <- match.arg(method)
#'
#'  # wrapper for random generation of AR1 series
#'  redNoise <- function(a1, n) {c(arima.sim(list(ar = a1), n))}
#'
#'  # MC sampling of p values
#'  pval <- array(dim = c(2, nr))
#'  for (i in 1 : nr) {
#'
#'    x <- redNoise(a1.x, nobs)
#'    if (method == "two.sample") y <- redNoise(a1.y, nobs)
#'
#'    if (method == "one.sample") {
#'
#'      tmp1 <- t.test(x)
#'      tmp2 <- t_test(x, a1.x = a1.x)
#'
#'    } else {
#'
#'      tmp1 <- t.test(x, y)
#'      tmp2 <- t_test(x, y, a1.x = a1.x, a1.y = a1.y)
#'
#'    }
#'
#'    # count false positives
#'    if (tmp1$p.value <= 0.05) pval[1, i] <- 1 else pval[1, i] <- 0
#'    if (tmp2$p.value <= 0.05) pval[2, i] <- 1 else pval[2, i] <- 0
#'
#'  }
#'
#'  # return null hypothesis rejection rates (type I error)
#'  c(t.test.default = sum(pval[1, ]) / nr,
#'    t_test = sum(pval[2, ]) / nr)
#'
#' }
#'
#' # The expectation is a rejection rate of around 0.05; in the following
#' # examples we see that the default t.test shows higher than expected
#' # rejection rates for autocorrelated data while the t_test remedies this
#' # issue.
#'
#' estimateRejectionRate(1000, 1000, 0.5)
#' estimateRejectionRate(1000, 1000, 0.9)
#' estimateRejectionRate(1000, 1000, 0.5, method = "two.sample")
#' estimateRejectionRate(1000, 1000, 0.5, 0.7, method = "two.sample")
#' estimateRejectionRate(1000, 1000, 0.2, 0.9, method = "two.sample")
#'
#' @export
#'
t_test <- function (x, y = NULL, a1.x = 0, a1.y = 0,
    alternative = c("two.sided", "less", "greater"),
    mu = 0, paired = FALSE, var.equal = FALSE, conf.level = 0.95,  
    ...) {

  alternative <- match.arg(alternative)
  if (!missing(mu) && (length(mu) != 1 || is.na(mu))) 
    stop("'mu' must be a single number")
  if (!missing(conf.level) && (length(conf.level) != 1 ||
                               !is.finite(conf.level) || 
                               conf.level < 0 || conf.level > 1)) 
    stop("'conf.level' must be a single number between 0 and 1")
  if (!is.null(y)) {
    dname <- paste(deparse1(substitute(x)), "and", deparse1(substitute(y)))
    if (paired) 
      xok <- yok <- stats::complete.cases(x, y)
    else {
      yok <- !is.na(y)
      xok <- !is.na(x)
    }
    y <- y[yok]
  }
  else {
    dname <- deparse1(substitute(x))
    if (paired) 
      stop("'y' is missing for paired test")
    xok <- !is.na(x)
    yok <- NULL
  }
  x <- x[xok]
  if (paired) {
    x <- x - y
    y <- NULL
  }
  ## tmuench, 2023/11/27 ->
  nx <- stattools::getEffectiveDOF(length(x), a1.x)
  ## <-
  mx <- mean(x)
  vx <- stats::var(x)
  if (is.null(y)) {
    if (nx < 2) 
      stop("not enough 'x' observations")
    df <- nx - 1
    stderr <- sqrt(vx/nx)
    if (stderr < 10 * .Machine$double.eps * abs(mx)) 
      stop("data are essentially constant")
    tstat <- (mx - mu)/stderr
    method <- if (paired) 
                "Paired t-test"
              else "One Sample t-test"
    estimate <- stats::setNames(mx, if (paired) 
                               "mean difference"
                             else "mean of x")
  }
  else {
    ## tmuench, 2023/11/27 ->
    ny <- stattools::getEffectiveDOF(length(y), a1.y)
    ## <-
    if (nx < 1 || (!var.equal && nx < 2)) 
      stop("not enough 'x' observations")
    if (ny < 1 || (!var.equal && ny < 2)) 
      stop("not enough 'y' observations")
    if (var.equal && nx + ny < 3) 
      stop("not enough observations")
    my <- mean(y)
    vy <- stats::var(y)
    method <- paste(if (!var.equal) 
                      "Welch", "Two Sample t-test")
    estimate <- c(mx, my)
    names(estimate) <- c("mean of x", "mean of y")
    if (var.equal) {
      df <- nx + ny - 2
      v <- 0
      if (nx > 1) 
        v <- v + (nx - 1) * vx
      if (ny > 1) 
        v <- v + (ny - 1) * vy
      v <- v/df
      stderr <- sqrt(v * (1/nx + 1/ny))
    }
    else {
      stderrx <- sqrt(vx/nx)
      stderry <- sqrt(vy/ny)
      stderr <- sqrt(stderrx^2 + stderry^2)
      df <- stderr^4/(stderrx^4/(nx - 1) + stderry^4/(ny - 
                                                      1))
    }
    if (stderr < 10 * .Machine$double.eps * max(abs(mx), 
                                                abs(my))) 
      stop("data are essentially constant")
    tstat <- (mx - my - mu)/stderr
  }
  if (alternative == "less") {
    pval <- stats::pt(tstat, df)
    cint <- c(-Inf, tstat + stats::qt(conf.level, df))
  }
  else if (alternative == "greater") {
    pval <- stats::pt(tstat, df, lower.tail = FALSE)
    cint <- c(tstat - stats::qt(conf.level, df), Inf)
  }
  else {
    pval <- 2 * stats::pt(-abs(tstat), df)
    alpha <- 1 - conf.level
    cint <- stats::qt(1 - alpha/2, df)
    cint <- tstat + c(-cint, cint)
  }
  cint <- mu + cint * stderr
  names(tstat) <- "t"
  names(df) <- "df"
  names(mu) <- if (paired) 
                 "mean difference"
               else if (!is.null(y)) 
                 "difference in means"
               else "mean"
  attr(cint, "conf.level") <- conf.level
  rval <- list(statistic = tstat, parameter = df, p.value = pval, 
               conf.int = cint, estimate = estimate, null.value = mu, 
               stderr = stderr, alternative = alternative, method = method, 
               data.name = dname)
  class(rval) <- "htest"
  rval
}

# Code adapted from base stats:::ks.test.default as of R version 4.3.1:
# GPL-2 | GPL-3
# Copyright (C) 2023 R Core Team

# Changes to original code by Thomas Muench in 2023/11/27:
# - introduced function parameters `a1.x` and `a1.y` to allow input of AR1
#   autocorrelation parameters for data x and y;
# - use these parameters to reset the number of x/y data points to the effective
#   number of degrees of freedom using Xu's formula neff = n * (1 - a).

#' Kolmogorov-Smirnov Test for autocorrelated data
#'
#' Perform a two-sample Kolmogorov-Smirnov test accounting for autocorrelation
#' of the input data. This function is a modified version of the base R's
#' default two-sample \code{\link[stats]{ks.test}}, which treats autocorrelation
#' after Xu (2013) by resetting the number of observations to \code{n * (1 -a)},
#' where \code{n} are the original number of observations and \code{a} is the
#' user-supplied estimate of the data's autocorrelation parameter at lag
#' one. Note that this code is adapted from the default method of the ks.test
#' function, thus it can only test whether the data are from the same or from
#' different distributions and not test the data against a given cumulative
#' distribution function, nor can it be used with class \code{formula}.
#'
#' @param x a numeric vector of data values.
#' @param y a numeric vector of data values.
#' @param a1.x autocorrelation parameter at lag one of the data in \code{x}.
#' @param a1.y autocorrelation parameter at lag one of the data in \code{y}.
#' @inheritParams stats::ks.test
#' @return see \code{?ks.test}.
#' @seealso \code{\link[stats]{ks.test}}
#' @source \code{stats:::ks.test.default}
#' @references
#'   Xu, X.: Methods in Hypothesis Testing, Markov Chain Monte Carlo and
#'   Neuroimaging Data Analysis, PhD thesis, Harvard University, Cambridge,
#'   Massachusetts, \url{http://nrs.harvard.edu/urn-3:HUL.InstRepos:11108711},
#'   2013.
#' @author R Core Team, Thomas Münch
#' @examples
#'
#' # Check the performance of the test on autocorrelated data
#'
#' # This function Monte Carlo samples autocorrelated data series from an AR1
#' # process and runs the default ks test as well as the stattools ks_test to
#' # estimate the tests' rejection rates, i.e. the relative number of false
#' # positives where the tests reject the null hypothesis
#' estimateRejectionRate <- function(nobs, nr, a1.x, a1.y = a1.x) {
#'
#'  # wrapper for random generation of AR1 series
#'  redNoise <- function(a1, n) {c(arima.sim(list(ar = a1), n))}
#'
#'  # MC sampling of p values
#'  pval <- array(dim = c(2, nr))
#'  for (i in 1 : nr) {
#'
#'    x <- redNoise(a1.x, nobs)
#'    y <- redNoise(a1.y, nobs)
#'
#'    tmp1 <- ks.test(x, y)
#'    tmp2 <- ks_test(x, y, a1.x = a1.x, a1.y = a1.y)
#'
#'    # count false positives
#'    if (tmp1$p.value <= 0.05) pval[1, i] <- 1 else pval[1, i] <- 0
#'    if (tmp2$p.value <= 0.05) pval[2, i] <- 1 else pval[2, i] <- 0
#'
#'  }
#'
#'  # return null hypothesis rejection rates (type I error)
#'  c(ks.test.default = sum(pval[1, ]) / nr,
#'    ks_test = sum(pval[2, ]) / nr)
#'
#' }
#'
#' # The expectation is a rejection rate of around 0.05; in the following
#' # examples we see that the default ks.test shows higher than expected
#' # rejection rates for autocorrelated data while the ks_test remedies this
#' # issue for identical autocorrelation...
#'
#' estimateRejectionRate(1000, 1000, 0.5)
#' estimateRejectionRate(1000, 1000, 0.9)
#'
#' # ..., however, it shows rejection rates > 0.05 if the autoccorelation is
#' # very different, for which the reason is unclear at the moment, with better
#' # rejection rates for small sample sizes
#'
#' estimateRejectionRate(1000, 1000, 0.5, 0.6)
#' estimateRejectionRate(1000, 1000, 0.5, 0.8)
#' estimateRejectionRate(99, 1000, 0.5, 0.8)
#'
#' @export
#'
ks_test <- function (x, y, a1.x = 0, a1.y = 0,
                     alternative = c("two.sided", "less", "greater"), 
                     exact = NULL, simulate.p.value = FALSE, B = 2000) {

  alternative <- match.arg(alternative)
  DNAME <- paste(deparse1(substitute(x)), "and", deparse1(substitute(y)))
  x <- x[!is.na(x)]
  n.x <- length(x)
  if (n.x < 1L) 
    stop("not enough 'x' data")
  PVAL <- NULL
  if (is.ordered(y)) 
    y <- unclass(y)
  y <- y[!is.na(y)]
  n.y <- length(y)
  if (n.y < 1L) 
    stop("not enough 'y' data")
  n.x <- as.double(n.x)
  if (is.null(exact)) 
    exact <- (n.x * n.y < 10000)
  if (!simulate.p.value) {
    METHOD <- paste(c("Asymptotic", "Exact")[exact + 1L],
                    "two-sample Kolmogorov-Smirnov test")
  }
  else {
    METHOD <- "Monte-Carlo two-sample Kolmogorov-Smirnov test"
  }
  TIES <- FALSE
  w <- c(x, y)
  z <- cumsum(ifelse(order(w) <= n.x, 1/n.x, -1/n.y))
  if (length(unique(w)) < (n.x + n.y)) {
    z <- z[c(which(diff(sort(w)) != 0), n.x + n.y)]
    TIES <- TRUE
    if (!exact && !simulate.p.value) 
      warning("p-value will be approximate in the presence of ties")
  }
  STATISTIC <- switch(alternative, two.sided = max(abs(z)), 
                      greater = max(z), less = -min(z))
  nm_alternative <- switch(alternative, two.sided = "two-sided", 
                           less = "the CDF of x lies below that of y",
                           greater = "the CDF of x lies above that of y")
  z <- NULL
  if (TIES) 
    z <- w
  ## tmuench, 2023/11/27 ->
  n.x.eff <- n.x * (1 - a1.x)
  n.y.eff <- n.y * (1 - a1.y)
  PVAL <- switch(alternative,
                 two.sided = stats::psmirnov(STATISTIC,
                                             sizes = c(n.x.eff, n.y.eff),
                                             z = w, exact = exact,
                                             simulate = simulate.p.value, B = B,
                                             lower.tail = FALSE),
                 less = stats::psmirnov(STATISTIC,
                                        sizes = c(n.x.eff, n.y.eff), ,
                                        z = w, exact = exact,
                                        simulate = simulate.p.value, B = B,
                                 two.sided = FALSE, lower.tail = FALSE),
                 greater = stats::psmirnov(STATISTIC,
                                           sizes = c(n.x.eff, n.y.eff), ,
                                           z = w, exact = exact,
                                           simulate = simulate.p.value, B = B,
                                           two.sided = FALSE,
                                           lower.tail = FALSE))
  ## <-
  if (simulate.p.value)
    PVAL <- (1 + (PVAL * B))/(B + 1)

  names(STATISTIC) <- switch(alternative, two.sided = "D", 
                             greater = "D^+", less = "D^-")
  PVAL <- min(1, max(0, PVAL))
  RVAL <- list(statistic = STATISTIC, p.value = PVAL,
               alternative = nm_alternative, method = METHOD, data.name = DNAME,
               data = list(x = x, y = y), exact = exact)
  class(RVAL) <- c("ks.test", "htest")
  return(RVAL)
}
